Upgrade toolchain to nightly-2024-04-15

celinval opened the pull request:
Related changes:
  - https://github.com/rust-lang/rust/pull/118310:
    Add `Ord::cmp` for primitives as a `BinOp` in MIR
  - https://github.com/rust-lang/rust/pull/120131:
    Add support to `Pat` pattern type
  - https://github.com/rust-lang/rust/pull/122935:
    Rename CastKind::PointerWithExposedProvenance
  - https://github.com/rust-lang/rust/pull/123097:
    Adapt to changes to local_def_path_hash_to_def_id

Resolves #3130, #3142

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

celinval commented:
Note to self: Add a test to Pat to ensure we capture invalid valid generated during a transmute.

tautschnig reviewed (commented):
For my own learning: the changes in https://github.com/model-checking/kani/pull/3144/commits/1abfa0fd7792f84eaff5aa201627739771d57a2e ("Fix BinOp implementation") seem much bigger than what I initially had for the Cmp operator. Could you please explain a bit?

celinval commented on cprover_bindings/src/goto_program/expr.rs:
```diff
@@ -389,6 +389,7 @@ impl Expr {
             source.is_integer() || source.is_pointer() || source.is_bool()
         } else if target.is_integer() {
             source.is_c_bool()
+                || source.is_bool()
```
So do I need to cast to c_bool first and then to to integer? Because that seems to be allowed.

tautschnig replied:
Oh, my apologies, I had failed to consider the context in which this check is being made. It seems we are just concerned with the bit representation here, so this is actually fine.

celinval commented:
> For my own learning: the changes in [1abfa0f](https://github.com/model-checking/kani/commit/1abfa0fd7792f84eaff5aa201627739771d57a2e) ("Fix BinOp implementation") seem much bigger than what I initially had for the Cmp operator. Could you please explain a bit?

I added a comment to the code, which explains why this is slightly more complex. The result of a comparison is not an integer, but an enumeration. So Kani would crash with the previous implementation when it would try to assign the comparison expression due to mismatched type.

Part of the logic I added is to guarantee that the transmute is safe, but it is not really necessary. But I check that the enumeration layout matches what we expect, which is an i8 that goes from -1 to 1.

celinval commented:
I'm merging this PR despite the issue with the benchmark job, since it is blocked due to a bug in the benchcomp tool itself. It is failing to generate the comparison tables since the base commit is currently broken due to #3138
